masschallenge/impact-api
# [AC-6592] Application assignments should stale out over time

To test: 
Testing the allocator is difficult. Here's a way to see it working for a simple case.

- go to judging round 113 and set it to active, end date in future
- go to criterion option specs for the round:
http://localhost:8181/admin/accelerator/criterionoptionspec/?criterion__judging_round__id__exact=113
- set all criteria which are not "reads" to 0 weight, 0 count

- set up db so that only applications with a single assignment and no reads are live.
Something like:

jr = JudgingRound.objects.get(pk=133)
jr.program.cycle.applications.update(application_status="complete")
apps = jr.program.cycle.applications.annotate(read_count=Count("judgeapplicationfeedback"), assignment_count=Count("applicationpanelassignment")).filter(read_count=0, assignment_count=1)
apps.update(application_type=jr.application_type)
apps.update(application_status="assigned")

# get the assignments for these apps:
apas = ApplicationPanelAssignment.objects.filter(application__in=apps)

# set the created_at for the apas to random times in the last wee
now = utc.localize(datetime.now())
seven_days = 7 * 24 * 60 * 60
for apa in apas:
    apa.created_at = now - timedelta(seconds=random(seven_days)
    apa.save()

# list the applications sorted by apa.created_at:
apas.order_by("created_at").values_list("application__startup__organization__name")

- log in as a judge for this round, such as 32-user@example.com
- go to /panels and request apps (will need to have `make run-all-servers` for this to work)
- should see that the least recently assigned apps are assigned to this judge
